website/content/docs/internals/replication.mdx

```diff
@@ -142,6 +142,73 @@ Vault uses the merkle index as the source of truth, allowing the WAL streams to
+
+When a cluster is enabled as replication primary, it persists a cluster
+definition to storage, under `core/cluster/replicated/info` or
+`core/cluster/replicated-dr/info`.  An optional field of the cluster definition
+is `primary_cluster_addr`, which may be provided in the enable request.
+
+Perf standbys regularly issue heartbeat RPC requests to the active node, and
+one of the arguments to the RPC is the local node's `cluster_addr`.  The
+primary active node retains these cluster addresses received from its peers in
+an in-memory cache named `clusterPeerClusterAddrsCache` with a 15s expiry time.
+
+### Cluster addresses on the secondary
+
+When a secondary is enabled, its replication activation token (obtained from
+the primary) includes a `primary_cluster_addr` field.  This is taken from the
+persisted cluster definition created when the primary was enabled, or if no
+`primary_cluster_addr` was provided at that time, the token contains the
+`cluster_addr` of the current active node in the primary at the time the
+activation token is created.
+
+The secondary persists its own version of the cluster definition to storage,
+again under `core/cluster/replicated/info` or
+`core/cluster/replicated-dr/info`.  Here the `primary_cluster_addr` field is
+the one obtained from the activation token.
+
+The secondary active node regularly issues heartbeat RPC requests to the
+primary active node.  In response to these, the primary returns a response
+which includes a `ClusterAddrs` field, comprising the contents of its
+`clusterPeerClusterAddrsCache` plus the current active node's `cluster_addr`.
+The secondary uses the response to both to update its in-memory record of
+`known_primary_cluster_addrs`, and in addition it persists them to storage
+under `core/primary-addrs/dr` or `core/primary-addrs/perf`.  When this happens
+it logs the line `"replication: successful heartbeat"`, which includes the
+`ClusterAddrs` value obtained in the response.
+
+### Secondary RPC address resolution
+
+gRPC is given a list of target addresses for it to use in performing RPC
+requests.  gRPC will discover that perf standbys can't service most RPCs, and
+will quickly weed out all but the active node cluster address.  If the primary
+active node changes, gRPC will learn that its address is no longer viable and
+will automatically fail over to the new active node, assuming it's one of the
+known target addresses.
+
+The secondary runs a background resolver goroutine that, every few seconds,
+builds the gRPC target list of addresses for the primary.  Its output is logged
+at Trace level as `"loaded addresses"`.
+
+To build the primary cluster address list, the resolver goroutine normally
+simply concatenates `known_primary_cluster_addrs` with the
+`primary_cluster_addr` in the cluster definition in storage.
+
+There are two exceptions to that normal behaviour: the first time the goroutine
+goes through its loop, and when gRPC asks for a forced ResolveNow, which seems
```

Comment:
Instead of `which seems to happen`, can we say `which happens`?

Reply:
I used "seems" because that's what I took from the code comment Brian wrote around the code doing the Resolve.  I agree it's not ideal in docs, but I'm also unhappy asserting something we're not sure about.  I'll think on it a bit, maybe see if I can find any gRPC docs that clarify things.

Reply:
Nothing definitive, but I've talked myself into removing the "seems to" qualifier.